[JENKINS-26580] Activate JNLP3 support

kohsuke opened the pull request:
[JENKINS-26580](https://issues.jenkins-ci.org/browse/JENKINS-26580)

Follow up to https://github.com/jenkinsci/remoting/pull/69 to activate JNLP3 support on the server side

akshayabd committed:
[JENKINS-26580] Initial implementation of JNLP3-connect protocol

akshayabd committed:
[JENKINS-26580] Updated implementation of Jnlp3 protocol

kohsuke committed:
Moved some of the server-side logic to remoting

This way, the inner-working of the protocol can be kept private within
the remoting module, which increases a modularity.

kohsuke committed:
JnlpAgentReceiver needs to be adjusted for the new type

Given that there's only one caller, I think handling AbstractMethodError
is better than doing the compatibility method ping pong

kohsuke committed:
Merge remote-tracking branch 'origin/master' into jnlp3-protocol

Conflicts:
	core/src/main/java/jenkins/slaves/JnlpAgentReceiver.java
	pom.xml

kohsuke commented:
cc @akshayabd @oleg-nenashev @jglick 

Merging this requires a corresponding remoting release

oleg-nenashev commented on core/src/main/java/jenkins/slaves/JnlpAgentReceiver.java:
```diff
@@ -55,7 +57,16 @@
      *      Any exception thrown from this method will fatally terminate the connection.
      */
-    public abstract boolean handle(String name, JnlpSlaveHandshake handshake) throws IOException, InterruptedException;
+    public abstract boolean handle(String name, JnlpServerHandshake handshake) throws IOException, InterruptedException;
```
It still breaks the binary compatibility, because external classes won't implement the new abstract method. Maybe NIT in the case of remoting

oleg-nenashev replied:
BTW the invocation about catches the AbstractMethodError

oleg-nenashev commented on core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol3.java:
```diff
@@ -0,0 +1,101 @@
+ *
+ * <p>@see {@link org.jenkinsci.remoting.engine.JnlpProtocol3} for more details.
+ *
+ * @author Akshay Dayal
```
<code>@since</code> would be very useful in this case

oleg-nenashev commented:
LGTM in general.

amuniz commented on core/src/main/java/jenkins/slaves/JnlpAgentReceiver.java:
```diff
@@ -55,7 +57,16 @@
+     * @deprecated
+     *      Use {@link #handle(String, JnlpServerHandshake)}
+     */
+    public boolean handle(String name, JnlpSlaveHandshake handshake) throws IOException, InterruptedException {
```
Missing `@Deprecated` - given the javadoc.

kohsuke replied:
One or the other is sufficient.

amuniz commented on core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol3.java:
```diff
@@ -0,0 +1,101 @@
+        protected String getNodeSecret(String nodeName) throws Failure {
+            computer = (SlaveComputer) Jenkins.getInstance().getComputer(nodeName);
+            if (computer == null) {
+                throw new Failure("Slave trying to register for invalid node: " + nodeName);
```
Agent

kohsuke committed:
Fixing a compilation problem

kohsuke committed:
Java7 code transformation

kohsuke committed:
Log cannot be opened twice so it needs to be kept around.

kohsuke commented:
My plane is about to leave so I can't check this now but does this code correctly call ChannelConfigurator

kohsuke committed:
Terminology fix

jglick commented on core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol2.java:
```diff
@@ -54,8 +55,13 @@ protected void run() throws IOException, InterruptedException {
+                    if (recv.handle(nodeName,this))
+                        return;
+                } catch (AbstractMethodError e) {
+                    if (recv.handle(nodeName,new JnlpSlaveHandshake(this)))
```
Looks alarming. Why not just retain the original method signature, deprecate it, and delegate?

kohsuke replied:
This was much shorter than the ping-pong and loop detection, plus it has the benefit of abstract method clearly marked as such.

jglick commented:
> does this code correctly call `ChannelConfigurator`

The fix of SECURITY-206 includes an assertion in `JnlpAccessWithSecuredHudsonTest.testServiceUsingDirectSecret`. If the agent and master automatically agree to use the v3 protocol, I suppose the existing test should cover this.

jglick commented:
> `Could not resolve dependencies for project org.jenkins-ci.main:cli:jar:1.647-SNAPSHOT: Could not find artifact org.jenkins-ci.main:remoting:jar:2.54-SNAPSHOT in cloudbees-private-snapshot-repository`

You need to `deploy` a snapshot of the library to have this PR be tested.

kohsuke committed:
2.54 contains JNLP3 change

kohsuke commented:
Wow, Jesse predicted the exact test case that fails!!

kohsuke committed:
Need to call ChannelConfigurator

kohsuke committed:
When this gets merged let's record the version number

akshayabd commented on core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol3.java:
```diff
@@ -0,0 +1,112 @@
+ * @author Akshay Dayal
+ * @since 1.XXX
+ */
+@Extension
```
Initial manual testing looks good - the JNLP node connects with JNLP3 protocol and I can build a project on the node.

I think more stress testing needs to be done before JNLP3 is made the default protocol. In this initial release JNLP3 should only be available if the user explicitly asks for it. There are 3 approaches:

Master-side control
    1. Making master support JNLP3 through a flag
    2. Making master support JNLP3 through UI option
Node-side control
    3. Making node try JNLP3 first using flag

Option 1 can be done in "easy" ways I suppose, this class can check for a boolean system property, if its not set fail and let JNLP2 take over.
Option 2 is a lot of work and I would not favor it.
Option 3 is easy, but requires some extra work for users starting up nodes with JNLP3.

I'm ok with either option 1 or 3, prefer 1 slightly since I like controlling the behavior from the master side, not the node.

kohsuke replied:
OK, I'll ease this in by making this feature optional in the beginning for a while.

akshayabd replied:
I'm going to spend some time dusting out the stress-test suites that we used before and run them for JNLP3. Will probably have to modify them a bit to have significant "work" being done for each build instead of just "sleep 5". Will let you know when I'm done.

akshayabd commented on core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java:
```diff
@@ -27,7 +28,7 @@
 @Extension
 public class DefaultJnlpSlaveReceiver extends JnlpAgentReceiver {
```
Hmmm this is approach is slightly different than what I had in mind when you mentioned refactoring the remoting module. I was wondering if we could do-away with the server-side implementing JNLP version specific protocol handlers entirely? ie. The master would just make one call to the remoting library and the implementation there would try protocol 3, 2, 1.

This way we can add protocols to the remoting library without needing to update the master, and makes the remoting  library more re-usable and out-of-the-box ready.

Regardless this is fine since it works.

kohsuke replied:
Oh I see, that makes sense. I think we can take on that as a follow-up change.

kohsuke committed:
A/B test this feature in production

JNLP3 is activated now for 10% of users. Let's keep it like this for a
while and if no major issue occurs we should expose it to everyone.

kohsuke committed:
Merge remote-tracking branch 'origin/master' into jnlp3

Conflicts:
	pom.xml

kohsuke commented:
I discovered that the protocol doesn't gracefully handle a server that doesn't understand the protocol. Needs further tweak in remoting.

kohsuke committed:
Pick up a new version of remoting

See https://github.com/jenkinsci/remoting/pull/78

kohsuke commented:
Can somebody help me understand why the PR validation build doesn't pick up http://repo.jenkins-ci.org/snapshots/org/jenkins-ci/main/remoting/2.56-SNAPSHOT/ ?

kohsuke commented:
FWIW locally it passes all the tests.

aheritier commented:
@kohsuke https://repository-jenkins.forge.cloudbees.com/snapshot/org/jenkins-ci/main/remoting/ there is no 2.56-SNAPSHOT version here ....
Did you manually deployed it ?
I suppose you did and it was removed because Maven tries to get a version 2.56-20160304.223037-1 thus it must have this version in its local repository cache. The problem is that it doesn't use it because it isn't available in the remote repository. It is the crappy feature added in Maven 3.x and which always reports a stupid error message that doesn't explain the error cc @stephenc 
Building with -U option (or repositories options in settings to always check snapshots) is safer/cleaner to enforce maven to try to get a new SNAPSHOt if available. It won't solve the issue that the SNAPSHOT was deleted in the remote. The problem can be bypassed (but it's crappy) and let maven use its local cached SNAPSHOT with the option

```
-llr,--legacy-local-repository         Use Maven 2 Legacy Local
                                        Repository behaviour, ie no use of
                                        _remote.repositories. Can also be
                                        activated by using
                                        -Dmaven.legacyLocalRepo=true
```

kohsuke commented:
@aheritier see the link I provided above, which is the official snapshot repo for the Jenkins project. Sounds like the issue is that we are not configuring our POMs to refer to that snapshot repository, though it makes me wonder how others have been doing this.

aheritier commented:
@kohsuke https://repository-jenkins.forge.cloudbees.com/snapshot/org/jenkins-ci/main/remoting/ is where the build is looking for snapshots instead of http://repo.jenkins-ci.org/snapshots/org/jenkins-ci/main/remoting

I agree that the snapshot repo http://repo.jenkins-ci.org/snapshots/org/jenkins-ci/main/remoting should be defined in the POM or in the maven settings

For now it's not defined in any POM
https://github.com/jenkinsci/pom/blob/master/pom.xml#L77
https://github.com/jenkinsci/jenkins/blob/master/pom.xml#L111

Thus I imagine that people using SNAPSHOTs are defining the repo in the maven settings used on CI
There is a settings.xml config defined in the Config File Provider plugin called `JenkinsSettings` activating in a `jenkins` profile snapshots from http://repo.jenkins-ci.org/public/ but this one isn't used by the job. I don't know what are the rules and how is managed this instance. maybe @abayer @rtyler or @daniel-beck know ...

kohsuke committed:
Pick up a released version

kohsuke commented:
The test failure is a fluke

aheritier commented:
@kohsuke Great !!
I think that we should use for all jobs on https://jenkins.ci.cloudbees.com the same set of (maven) settings and to activate SNAPSHOTs for everyone (but I have no clue where to request it - INFRA ? - and who can/should take care of this)

akshayabd commented:
Great to see this merged in - thanks @kohsuke! I always recommend caution, I still haven't done enough scale testing to see what sort of [CPU] performance requirements the Ciphers have. Is there a smaller Jenkins cluster you can test this on before activating it on the main https://jenkins.ci.cloudbees.com?

stephenc commented:
Also, as I have pointed out, this is a non-NIO implementation and requires one reader thread per slave... I am working on a v4 protocol that uses a new SSLEngine aware NIOHub and hope to have that soon.

The consideration of @kohsuke and I is that allowing some encryption of the channel at the cost of threads is more important than waiting while I battle with SSLEngine to deliver a v4 protocol

Sent from my iPad

> On 9 Mar 2016, at 19:43, akshayabd <notifications@github.com> wrote:
> 
> Great to see this merged in - thanks @kohsuke! I always recommend caution, I still haven't done enough scale testing to see what sort of [CPU] performance requirements the Ciphers have. Is there a smaller Jenkins cluster you can test this on before activating it on the main https://jenkins.ci.cloudbees.com?
> 
> —
> Reply to this email directly or view it on GitHub.
>

akshayabd commented:
@stephenc Yes good catch, I had not considered the NIOHub issue. Performance-wise it would probably make a significant difference.
